examples/fit_with_bounds.py

```diff
@@ -0,0 +1,143 @@
+Eventually, we want to use parts of this in the base library, so we don't have
+to retype them every time.
+
+"""
+import matplotlib.pyplot as plt
+from numpy import exp, linspace, pi, random, sign, sin
+import sympy as sp
+from lmfit import Parameters, minimize
+from lmfit.printfuncs import report_fit
+import numpy as np
+from pyspecdata import *
+
+init_logging(level="debug")
+np.random.seed(15816)
+# {{{ helper function(s)
+def gen_from_expr(expr, guesses={}):
+    """generate parameter descriptions and a numpy (lambda) function from a sympy expresssion
+
+    Parameters
+    ==========
+    expr: sympy expression
+    guesses: dict
+        dictionary of keyword arguments for guesses (value) or constraints
+        (min/max)
+
+    Returns
+    =======
+    pars: lmfit.Parameters
+    parameter_names: tuple
+        ordered list of the names of the arguments  to fn
+    fn: function
+        the fit function
+    """
+    # {{{ decide which symbols are parameters vs. variables
+    all_symbols = expr.atoms(sp.Symbol)
+    axis_names = set([sp.Symbol(j) for j in empty_data.dimlabels])
+    variable_symbols = axis_names & all_symbols
+    parameter_symbols = all_symbols - variable_symbols
+    variable_symbols = tuple(variable_symbols)
+    variable_names = tuple([str(j) for j in variable_symbols])
+    parameter_symbols = tuple(parameter_symbols)
+    parameter_names = tuple([str(j) for j in parameter_symbols])
+    logger.info(
+        strm(
+            "all symbols are",
+            all_symbols,
+            "axis names are",
+            axis_names,
+            "variable names are",
+            variable_names,
+            "parameter names are",
+            parameter_names,
+        )
+    )
+    # }}}
+    pars = Parameters()
+    for this_name in parameter_names:
+        kwargs = {}
+        if this_name in guesses.keys():
+            logger.info(strm("applying bounds for", this_name))
+            kwargs.update(guesses[str(this_name)])
+        pars.add(this_name, **kwargs)
+    for j in pars:
+        logger.info(strm("fit param ---", j))
+    logger.info(strm(pars))
+    fn = lambdify(
+        variable_symbols + parameter_symbols,
+        expr,
+        modules=[{"ImmutableMatrix": np.ndarray}, "numpy", "scipy"],
+    )
+    return pars, parameter_names, fn
+
+
+# }}}
+# {{{ creating fake data
+true_values = {"A": 14.0, "period": 5.4321, "shift": 0.12345, "decay": 0.01000}
+p_true = Parameters()
+for k, v in true_values.items():
+    p_true.add(k, value=v)
+logger.info(strm("p_true is:", p_true))
+x_vals = linspace(0, 250, 1500)
+empty_data = nddata(x_vals, "x").copy(data=False)
+# }}}
+# {{{making sympy expression
+A, shift, period, decay, x = sp.symbols("A shift period decay x")
+expr = A * sp.sin(shift + x / period) * sp.exp(-((x * decay) ** 2))
+# seems likely that Parameters is an ordered list, in which case, we don't need
+# parameter_names -- **however** we need to check the documentation to see that
+# this is true
```

Comment:
@guinneak198 please do this!

Reply:
(check the lmfit documentation to see if Parameters is an ordered list -- if it is, you don't need to return `parameter_names` below)

Reply:
so, actually I went ahead and checked -- it is:

> In lmfit, this one-dimensional array is replaced by a Parameters object, which works as an ordered dictionary of Parameter objects with a few additional features and methods.